### PR TITLE
Initial work on generic models for pydantic v2

### PR DESCRIPTION
**Pull request by dmontagu**
Relevant issue: https://github.com/pydantic/pydantic/issues/4665

I have opened this as a draft pull request to get some initial feedback on:
* The inclusion of `typevars_map` as an argument to various schema-building functions
* The addition of some generics-related methods to `BaseModel`
* Moving some of the existing generics functionality to `pydantic.main` and some to `pydantic.generics_utils`. (Part of the motivation of creating a new file was to keep the PR small/easy-to-review for now. I am imagining that this file will eventually replace the current `pydantic.generics`.)
* The details of `pydantic.generics_utils.create_generic_submodel` and `BaseModel.__class_getitem__`

I am sure that there are still a variety of edge cases that are better-handled in pydantic v1 because I haven't tested them (e.g., anything depending on `_assigned_parameters`, which I haven't needed to migrate yet). 

If I get positive feedback on the approach in this PR, I will review those edge cases more thoroughly.

I have tried to bias toward copying less of the existing code when I didn't fully understand what it was used for, though I intend to check more thoroughly for any regressions soon (definitely before merging). Hopefully the mostly-disabled tests of `GenericModel` cover most of these..

### CODE REVIEW

File: pydantic/main.py [hunk 1]
```diff
@@ -592,6 +603,51 @@ def __repr_args__(self) -> _repr.ReprArgs:
...
+    def __concrete_name__(cls: type[Any], params: typing.Tuple[type[Any], ...]) -> str:
+        """Compute class name for child classes.
+
+        :param params: Tuple of types the class . Given a generic class
+            `Model` with 2 type variables and a concrete model `Model[str, int]`,
+            the value `(str, int)` would be passed to `params`.
+        :return: String representing a the new class where `params` are
+            passed to `cls` as type variables.
+
+        This method can be overridden to achieve a custom naming scheme for GenericModels.
```
Comment (dmontagu) [inline pydantic/main.py:645]: Obviously need to update these docs, but I wasn't sure how you were going to feel about keeping `__concrete_name__` as a method on `BaseModel`. Having the ability to modify the name-generation logic feels like a significant benefit, I just feel a bit averse to polluting the `BaseModel` namespace.

File: pydantic/_internal/_generate_schema.py [hunk 2]
```diff
@@ -117,6 +124,26 @@ def generate_schema(self, obj: Any) -> core_schema.CoreSchema:  # noqa: C901
             if issubclass(obj, dict):
                 return self._dict_subclass_schema(obj)
             # probably need to take care of other subclasses here
+        elif isinstance(obj, typing.TypeVar):
+            # TODO: Add a check for recursion depth exception and an explanatory re-raise
```
Comment (dmontagu) [inline pydantic/_internal/_generate_schema.py:128]: (note to self -- need to delete this comment; the exception handling has already been added below)

File: pydantic/generics_utils.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+from typing import Any
+
+TypeVarType = Any  # since mypy doesn't allow the use of TypeVar as a type
+Parametrization = typing.Mapping[TypeVarType, type[Any]]
+
+if typing.TYPE_CHECKING:
+    from pydantic import BaseModel
+
+    class GenericBaseModel(BaseModel):
+        __parameters__: typing.ClassVar[typing.Tuple[TypeVarType, ...]]
```
Comment (dmontagu) [inline pydantic/generics_utils.py:13]: This type-checking-only `GenericBaseModel` class, along with the `is_generic_model` typeguard below, are just there to help address the fact that mypy doesn't like when I access `__parameters__` on a class that inherits from both `BaseModel` and `typing.Generic`. This approach seemed to handle it correctly with minimal mypy ignores.

File: pydantic/generics_utils.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        previous_caller_frame = sys._getframe(depth)
+    except ValueError as e:
+        raise RuntimeError('This function must be used inside another function') from e
+    except AttributeError:  # sys module does not have _getframe function, so there's nothing we can do about it
+        return None, False
+    frame_globals = previous_caller_frame.f_globals
+    return frame_globals.get('__name__'), previous_caller_frame.f_locals is frame_globals
+
+
+def is_generic_model(x: type['BaseModel']) -> typing.TypeGuard[type['GenericBaseModel']]:
```
Comment (dmontagu) [inline pydantic/generics_utils.py:57]: Note that, where used, this type guard _is required_ at runtime (for the sake of generating an appropriate error message if you try to parameterize a non-generic BaseModel). In particular, even though the type guard provides typing benefits, and even though the type it is "guarding" isn't even "real", the runtime overhead of the check is still required.

File: pydantic/main.py [hunk 5]
```diff
@@ -592,6 +603,51 @@ def __repr_args__(self) -> _repr.ReprArgs:
             if not k.startswith('_') and (k not in self.model_fields or self.model_fields[k].repr)
         ]
 
+    def __class_getitem__(cls, item: typing.Union[type[Any], typing.Tuple[type[Any], ...]]) -> type[Any]:
```
Comment (samuelcolvin) [inline pydantic/main.py:606]: i think you can use `|` union syntax here.

File: pydantic/main.py [hunk 6]
```diff
@@ -42,6 +43,10 @@
 # the `BaseModel` class, since that's defined immediately after the metaclass.
 _base_class_defined = False
 
+_generic_types_cache: _utils.LimitedDict[
+    typing.Tuple[type[Any], Any, typing.Tuple[Any, ...]], type[BaseModel]
```
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/main.py:47]: It might be worth looking at #3829 #4551 to check we don't introduce any of those problems again - the memory leak issue should be solved by `LimitedDict`, but the `TypeError` could still exist I guess.

File: pydantic/main.py [hunk 7]
```diff
@@ -19,6 +19,7 @@
 from .config import BaseConfig, Extra, build_config, inherit_config
 from .errors import PydanticUserError
 from .fields import Field, FieldInfo, ModelPrivateAttr
+from .generics_utils import create_generic_submodel, is_generic_model
```
Comment (samuelcolvin) [inline pydantic/main.py:22]: to avoid private methods being publicly importable (e.g. with `from pydantic.main import create_generic_submodel`, better to add `_generics` to the `from ._internal import ...` line.

File: pydantic/main.py [hunk 8]
```diff
@@ -592,6 +603,51 @@ def __repr_args__(self) -> _repr.ReprArgs:
...
+    def __class_getitem__(cls, item: typing.Union[type[Any], typing.Tuple[type[Any], ...]]) -> type[Any]:
+        def _cache_key(_params: Any) -> typing.Tuple[type[Any], Any, typing.Tuple[Any, ...]]:
+            return cls, _params, typing_extensions.get_args(_params)
+
+        cached = _generic_types_cache.get(_cache_key(item))
+        if cached is not None:
+            return cached
+
+        if not isinstance(item, tuple):
+            item = (item,)
```
Comment (samuelcolvin) [inline pydantic/main.py:615]: is it worth having a more descriptive name, e.g.
```suggestion
            typevar_values = (item,)
        else:
            typevar_values = item
```
???

File: pydantic/main.py [hunk 9]
```diff
@@ -592,6 +603,51 @@ def __repr_args__(self) -> _repr.ReprArgs:
...
+
+        if not is_generic_model(cls):
```
Comment (samuelcolvin) [inline pydantic/main.py:617]: should this come first in the function.

File: pydantic/main.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic/main.py:645]: new rule for v2 to try and be better python citizens_ we want all dunder names to start with `__pydantic` (or at least include pydanitc, e.g. `__get_pydantic...`).

Should we use `__pydantic_concrete_name__` here?

(I assume `__concrete_name__` is not defined in the stdlib, or a pepe somewhere?)

Or, since this is not really private or magic, maybe we should just use `model_concrete_name`???

Comment (samuelcolvin) [reply pydantic/main.py:645]: I I think I prefer a public name, e.g. `model_concrete_name`.

File: pydantic/main.py [hunk 10]
```diff
@@ -592,6 +603,51 @@ def __repr_args__(self) -> _repr.ReprArgs:
...
+
+        :param params: Tuple of types the class . Given a generic class
+            `Model` with 2 type variables and a concrete model `Model[str, int]`,
+            the value `(str, int)` would be passed to `params`.
+        :return: String representing a the new class where `params` are
+            passed to `cls` as type variables.
+
+        This method can be overridden to achieve a custom naming scheme for GenericModels.
+        """
+        param_names = [_repr.display_as_type(param) for param in params]
```
Comment (samuelcolvin) [inline pydantic/main.py:647]: if this is only used on generics, should we raise an error if the model is not generic?

**Review by samuelcolvin — COMMENTED** [review]
Overall looks great.

Quite satisfying to see that the new core/core-schema approach has simplified this a lot, at least that's how I see it?!

Might be worth looking through the `1.10.X-fixes` branch and see if there are any bug fixes there that should be included here? E.g. #4551 mentioned below, you might just copy any new tests and check they pass.

Please update.

File: pydantic/main.py (hunk 1 above)
Comment (dmontagu) [reply pydantic/main.py:645]: @samuelcolvin I'm changing it to `model_concrete_name` for now, but one other alternative I'd like to propose: for public methods, using `model__` as a prefix, rather than `model_` (so `model__concrete_name` instead of `model_concrete_name`; similarly `model__json`, etc.).

This feels significantly less likely to run into collisions with user-desired fields (or methods), and to my eye almost *more* readable (as I can immediately tell it is a pydantic model method, since nothing else uses a format like that), with only one extra character. And it retains the various benefits of the method name being pseudo-public (e.g., IDEs including it at the top of autocomplete, etc.).

If you consider this an already-finalized decision, or just hate this proposal, let's just pretend this comment never happened :).

File: pydantic/_internal/_generate_schema.py [hunk 11]
```diff
@@ -117,6 +124,26 @@ def generate_schema(self, obj: Any) -> core_schema.CoreSchema:  # noqa: C901
...
+                            f'The maximum recursion depth was exceeded while generating the schema for a TypeVar. '
+                            f'This likely indicates a cycle in the TypeVar substitutions map ({self.typevars_map=}). '
+                            'This may be resolved by using the TypeVars in the same order as the original '
+                            'parameterization, or by using entirely new ones.'
+                        ) from exc
+            if obj.__bound__:
+                return self.generate_schema(obj.__bound__)
+            if obj.__constraints__:
+                return self._union_schema(typing.Union[obj.__constraints__])
+            return core_schema.AnySchema(type='any')
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/_internal/_generate_schema.py:146]: I did think about this; I spent some time thinking through the implications of disallowing model classes with free typevars from being instantiated at all, and ultimately I think this approach is preferable.

I couldn't think of any cases where an error was merited that mypy wouldn't be able to catch with an appropriate configuration if you wanted to be aggressive about your types (i.e., using generic models with implicit Any parameter). And it felt to me that the flexibility benefits of being able to work with unspecified generics was worth more than other forms of strictness here.

-----

The only pattern that I could imagine being problematic would be something along the following lines:
```python
class MyGenericModel(BaseModel, Generic[T]):
    my_field: T

MyConcreteModel = MyGenericModel[MyType]
```
If there was a non-pathological example where `MyGenericModel(my_field=instance_of_my_type).my_field` could end up different than `MyConcreteModel(my_field=instance_of_my_type).my_field` due to (again, non-pathological) validator usage, etc., I think that would be concerning. I couldn't think of anything that would cause this though without going out of our way (e.g., by having non-idempotent validators or something like that).

File: pydantic/main.py (hunk 8 above)
Comment (dmontagu) [reply pydantic/main.py:615]: yeah, I'm in favor. I noticed that the `__class_getitem__` method is defined on `GenericModel` using `params` as the name of the argument; I renamed it to `item` here following the documentation of `__class_getitem__` but just realized that I think we used `params` as the name in `GenericModel` because that is what it is called in `typing.Generic`. I checked and confirmed that the function `__class_getitem__` is called using an arg, rather than a kwarg, so I was thinking we could maybe just call the argument to the `__class_getitem__` function `typevar_values`.

I think that would be my first preference, since you can then see the intent as soon as you see the function signature, but I could just use your suggestion here if you prefer. (I do think either way is an improvement.)

File: pydantic/main.py (hunk 10 above)
Comment (dmontagu) [reply pydantic/main.py:647]: I'm okay either way; I can't really imagine a circumstance where you'd want to use it for non-generic-models, if someone found a use case where it was worth using on non-generic models for some reason, they could always override the method to drop the check for generic and open an issue requesting it work for non-generic models..

File: pydantic/main.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic/main.py]: let's still with `model_concrete_name` here for consistency.

Happy to consider `model__` prefix, probably worth discussing on #4915 or in a new issue.

I dislike all proposals in this area, including my own - there's no elegant way around this problem.

File: pydantic/_internal/_generate_schema.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/_internal/_generate_schema.py]: okay, let's leave it.

File: pydantic/main.py (hunk 8 above)
Comment (samuelcolvin) [reply pydantic/main.py]: whichever you prefer, it's not a big issue.

**Comment by dmontagu** [comment]
@samuelcolvin while investigating https://github.com/pydantic/pydantic/issues/4146#issuecomment-1399677854, I noticed the following unintuitive (to me) behavior:

```python
from typing import Generic, TypeVar
from pydantic import BaseModel

DataT = TypeVar("DataT")

class MyGenericModel(BaseModel, Generic[DataT]):
    generic: list[DataT] | None


class A(BaseModel):
    wrapper: MyGenericModel[int]

m = MyGenericModel(generic=None)
print(A(wrapper=m))
```

```
Traceback (most recent call last):
  File "scratch_66.py", line 14, in <module>
    print(A(wrapper=m))
          ^^^^^^^^^^^^
  File "pydantic/pydantic/main.py", line 155, in __init__
    values, fields_set = __pydantic_self__.__pydantic_validator__.validate_python(data)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for MyGenericModel[int]
wrapper
  Input should be a valid dictionary [type=dict_type, input_value=MyGenericModel(generic=None), input_type=MyGenericModel]
```

This is happening because `m` here is an instance of `MyGenericModel`, which is a _superclass_ of `MyGenericModel[int]`. There's nothing specific to generics with this error — this is basically the same as doing:
```python
from pydantic import BaseModel


class Inner(BaseModel):
    x: int


class SubInner(Inner):
    pass


class Outer(BaseModel):
    wrapper: SubInner


Outer(wrapper=Inner(x=1))
```

However, I can't help but feel it's a little awkward that you _have_ to add the annotation to the type when initializing it to get it to pass validation as a field.

I'm not sure I have a great proposal for a resolution to this (though I think we might be able to do something where we re-validate the data for concrete parametrizations of generic models, even if we don't do that for more general model superclasses). I don't think it needs to be resolved now (or possibly ever), as things are still plenty functional if you have to use the right parametrization when instantiating the class, it just feels a little awkward since you don't need to do that for non-pydantic generic classes.

Curious if you have any thoughts.

**Comment by samuelcolvin** [comment]
Very easy to add a flag to either config or a model to skip the isinstancd check - just run validation against `.__dict__` and see what happens.

With this nascent flag set, any class instance with the right fields would pass, maybe that's okay.

The problem is when to enable it? I guess we could enable it for generics by default?

We could also do some kind of "is super class" check, which sounds weird but could be useful.

**Comment by dmontagu** [comment]
> We could also do some kind of "is super class" check, which sounds weird but could be useful.

While addressing some of the more complex tests I realized I was running into some awkwardnesses with the class hierarchy being generated (M[int, T][str] would be two "layers" down from M, whereas M[int, str] is only one; M[int, T][str] was a different class than M[T, str][int] was a different class than M[int, str]).

To handle this, I made some more extensive changes to the implementation (adding fields like `__pydantic_generic_parent__` to hold a reference to the last class created the "normal" way, which then gets reparametrized using the full set of arguments when you do partial parametrizations). This means that no matter how you go about doing a partial parametrization, you should get class equality if the parametrization is the same, and parametrization only adds one layer to the class hierarchy no matter how many times `__class_getitem__` is called (with unsubstituted TypeVars).

I think it may be possible to leverage this approach for a more (appropriately) limited version of the "superclass" check you are describing in a straightforward way -- basically allowing instances of the unparametrized parent class. But, I'm not totally sure that this is something that makes sense.

Also, just to be clear, I'm concerned about the increase in complexity I've added but I think it's going to be necessary to handle some cases of deeply nested models with the same TypeVar used at multiple levels in the hierarchy; the more complex edge cases for which bugfixes and, thankfully, tests have been added over the last couple years :). I am pushing my progress directly here but I want to be clear that right now I'm mostly trying to find a path forward that works to handle the various edge cases, and will work on simplifying / reducing the amount of new/unnecessary fields once I figure out _something_ that works.

**Comment by dmontagu** [comment]
One interesting thing I noticed -- when using a bounded or constrained TypeVar, we don't do class-creation-time validation of parameter compatibility.

In particular:
```python
from typing import TypeVar, Generic

from pydantic import BaseModel

T = TypeVar("T", bound=str)


class Model(BaseModel, Generic[T]):
    data: T

Model[int](data=1)  # no error due to TypeVar substitution with `int`, despite being a type system error
Model(data=1)  # error thanks to bound on TypeVar requiring an instance of `str`
```

This _would_ be caught by mypy, and I'm thinking we probably don't want to get into the business of being a runtime type **system** _in addition_ to being a runtime type validator, but just wanted to point this out.

-----

On a semi-related note, when it comes time to implement subclass checking, we _may_ want to use information about the covariance/contravariance of the TypeVars involved; I'm not sure whether 1.X does that currently (I'm assuming not, but I need to check how that all works). I guess it would come down to how hard it is to do runtime validation of subclass relationships of type parameters, writing that out makes me think probably not going to happen, but I'll look into it a bit.

**Comment by dmontagu** [comment]
On the note of compatibility with runtime subclass checking of generics:
```python
from typing import Sequence

issubclass(Sequence[int], Sequence[float])

# TypeError: Subscripted generics cannot be used with class and instance checks
```

@samuelcolvin I'm wondering if disallowing subclass checks entirely with parameterized generics might be smart for reducing the amount of complex behavior we need to support; notice PyCharm even shows an error when you try (this is the case even if you introduce an alias for `A[int]` below):

image

I can imagine some advantages to supporting this, but I'm wondering if it's smarter to leave this to the type system and use a discriminated-union-esque approach if you really need to differentiate between specific generic parametrizations at runtime.

**Comment by samuelcolvin** [comment]
We've agreed that we don't need subclass stuff to work since it doesn't work with typing generics.

**Comment by caniko** [comment]
I don't see [PEP646](https://peps.python.org/pep-0646/) or variadic generics, being mentioned in this draft. Are you planning on adding this later?

**Comment by dmontagu** [comment]
I've paused work on this temporarily while trying to work through JSON schema and some other smaller things, and trying to familiarize myself more with pydantic_core, where I want to move at least some of the generics logic to get around some of the more complex challenges (such as recursive generic models with long cycles).

That said, I am hoping to add support for variadic generics as part of this. If it turns out to be particularly complicated I may postpone that in favor of getting non-variadic generic models working, but I am (cautiously?) optimistic.

To be clear, the use case I have in mind would be something like:

```python
```

**Comment by caniko** [comment]
This is a fine example:

> ```python
> class MyGeneric(BaseModel, Generic[*Shape]):
>     x: int
>     y: tuple[*Shape]
> ```

Moreover, in `>=3.11`: 
```python
Tuple[str, *Tuple[int, int], bool] -> Tuple[str, int, int, bool]
```
This is an extension of the logic from your example, I suggest that you implement with unpacking in mind.

Also keep in mind that normal types like `TypeVar` instances can precede a `TypeVarTuple` instance:
```python
T = TypeVar('T')    # or NewType('T')
Y = TypeVar('Y')    # or NewType('Y')
Shape = TypeVarTuple('Shape')

class MyGeneric(BaseModel, Generic[T, Y, *Shape]):
    x: T
    y: tuple[*Shape]
    z: list[Y, *Shape]
```

**Comment by dmontagu** [comment]
Closing in favor of a new branch that I think is closer.
